### PR TITLE
fix(vow): simplify `watch` and promptly register reactions

### LINKED ISSUE #9002: `@agoric/vat-data/vow.js#watch` is not robust for long-lived promises
## Describe the bug

If a remote promise is supplied to `watch`, and the local vat restarts, then the watcher is not ever called even after the remote promise is settled.

## To Reproduce
1. Alice is a vat watching a non-vow promise decided by Bob, a remote vat.  Example: `watch(E(objectInBob).getLongLivedPromise(), durableWatcherInAlice)`
2. Alice is restarted/upgraded.
3. Bob fulfils the promise.
4. Neither the `onFulfilled` nor `onRejected` methods on `durableWatcherInAlice` are ever called.

## Expected behavior
Sometime shortly after Bob fulfils the promise, Alice's `durableWatcherInAlice.onFulfilled` callback will be called.

## Platform Environment
 - MacOS, Node 18.18.2
 - is there anything special/unusual about your platform? no
 - what version of the Agoric-SDK are you using? `@agoric/cosmic-swingset@0.41.3-1398-gd5ae0b8e5`

## Additional context
There is an incorrect use of `await p` in the implementation of `watch(p, watcher)` that defers the first call to `watchPromise(p, ...)`.  `watchPromise(p, ...)` must be called promptly so that the registration of the watcher can survive upgrades, regardless of how long it takes for `p` to settle.

## Screenshots
If applicable, add screenshots to help explain your problem, especially for UI interactions.

### PR DESCRIPTION
**Pull request by michaelfig**
refs: #8978
closes: #9002

## Description

Please review with whitespace changes turned off.

Existing `@agoric/vow` didn't accommodate using `watch` with promises that didn't quickly resolve to vows.

Overlooking this case led to some convoluted code.  This PR consolidates all the cases and makes for a much simpler `watch` implementation.

### Security Considerations

Fixes a correctness issue.

### Scaling Considerations

No significant changes.

### Documentation Considerations

n/a

### Testing Considerations

A failing unit test is included in the first commit, solved by the second commit.

### Upgrade Considerations

Correctly handles upgrade as advertised.

### CODE REVIEW

File: packages/vow/src/vow-utils.js [hunk 1]
```diff
@@ -38,21 +38,14 @@ const ALREADY_VOW = harden({});
  * @returns {Promise<U>}
  */
 export const unwrapPromise = async (specimenP, cb) => {
```
Comment (gibson042) [inline packages/vow/src/vow-utils.js]: Can you add a description of `unwrapPromise` to its doc comment? I don't have an intuitive understanding of what it's doing or what it's for, and the call sites don't help.

And likewise for the other functions in this package.

File: packages/vow/src/watch.js [hunk 2]
```diff
@@ -156,19 +145,11 @@ export const prepareWatch = (
     /** @type {import('./types.js').VowKit<TResult1 | TResult2>} */
     const { resolver, vow } = makeVowKit();
 
-    const { promiseWatcher, vowSetter } = makeWatcherKit(resolver, watcher);
+    // Create a promise watcher to track vows, retrying on disconnection.
```
Comment (gibson042) [inline packages/vow/src/watch.js]:
```suggestion
    // Create a promise watcher to track vows,
    // retrying upon rejection as controlled by `rejectionMeansRetry`.
```
Also, what are the precise semantics of "retry" here? I didn't track the introduction of vows as closely as I should have, but the general mechanics of the system at large are that restarting a vat causes the kernel to settle all promises yet to be decided by it with a rejection, prompting all sufficiently-informed subscribing vats to request a _replacement_ promise from the new incarnation using an approach that varies across code responsible for producing such a promise. I think the "vow" approach replaces such exported promises with exported durable objects, so is "retry" a matter of invoking a method on a vow to get a promise replacing the rejected promise from a previous incarnation?

File: packages/vow/src/watch.js [hunk 3]
```diff
@@ -74,58 +67,50 @@ const prepareWatcherKit = (zone, rejectionMeansRetry, watchVow) =>
...
+          this.state.vow = value;
+          watchVow(value, this.self);
+          return undefined;
+        }
+        this.state.watcher = undefined;
+        this.state.resolver = undefined;
+        if (!resolver) {
+          return undefined;
+        } else if (watcher) {
+          settle(resolver, watcher, 'onFulfilled', value);
```
Comment (gibson042) [inline packages/vow/src/watch.js]: I don't think I understand this block... can you explain why `watch(E(remotable).method(…), makeLongLivedWatcher(…))` should return a vow that resolves to results from the watcher's `onFulfilled`/`onRejected` method rather than resolving to match the eventual result from the remotable? Returning anything at all is a divergence from `watchPromise`, but I'm willing to accept that it makes sense for some purpose.

Comment (gibson042) [reply packages/vow/src/watch.js]: Also, while I'm on the topic of comparison (and understanding that this is not expected to be resolved in this PR but may prompt a followup issue), I note that [`watchPromise`](https://github.com/Agoric/agoric-sdk/blob/master/packages/swingset-liveslots/src/watchedPromises.js#L185) accepts a value to watch, a persistable (and in practice durable) watcher, and any remaining arguments to forward to the appropriate watcher method upon settlement—and this last feature is absent from the vow `watch` but heavily utilized in calls to `watchPromise` (e.g., Zoe uses [`watchPromise(E(adminNode).done(), watcher, instAdmin, adminNode)`](https://github.com/Agoric/agoric-sdk/blob/695c440c0f48a3591b15a43665682c5f1ebbad9d/packages/zoe/src/zoeService/startInstance.js#L101) to share a single watcher across all contract instances). Is the gap intentional?

File: packages/vow/test/test-watch.js [hunk 4]
```diff
@@ -50,6 +68,28 @@ const runTests = async t => {
...
+test('promise watcher', async t => {
+  const zone = makeHeapZone();
+  const { watch, when } = prepareVowTools(zone, {
+    rejectionMeansRetry: reason => reason === 'disconnected',
+  });
+
+  const pk = makePromiseKit();
+  pk.resolver.reject('disconnected');
+
+  const vow = watch(pk.promise, {
```
Comment (gibson042) [inline packages/vow/test/test-watch.js]: Let's please not define a `makePromiseKit` function with a different signature than the [same-named function in endo](https://github.com/endojs/endo/blob/a4a5a75a4200b471fb972061179d9c0b3b0c9f32/packages/promise-kit/index.js#L15-L29), even if only for testing. Another name would be fine, but my preference would be to just use the endo one.
```suggestion
  const { promise, reject } = makePromiseKit();
  reject('disconnected');

  const vow = watch(promise, {
```

File: packages/vow/test/test-watch.js [hunk 5]
```diff
@@ -50,6 +68,28 @@ const runTests = async t => {
...
+      return 'fulfilled';
+    },
+    onRejected(reason) {
+      t.is(reason, 'disconnected');
+      return `rejected ${reason}`;
+    },
+  });
 
-test('ack watcher - shim', runTests);
+  t.is(await when(vow), 'rejected disconnected');
```
Comment (gibson042) [inline packages/vow/test/test-watch.js:106]: What is this actually testing? I think we need some comments explaining why e.g. `rejectionMeansRetry: reason => reason === 'disconnected'` is not expected to have an effect here.

**Review by gibson042 — CHANGES_REQUESTED** [review]
We need an issue to describe the problem so this PR can be marked as closing it. @michaelfig do you have a good explanation for that?

Also, I think we're still missing test coverage of the situation possibly addressed by #8978, in which the vat watching a non-vow promise decided by a remote vat is itself restarted before that promise is fulfilled (e.g., `watch(E(remotable).getLongLivedPromise(), makeIncarnationCrossingDurableWatcher(…))`). And I'd like clarification from @iomekam that this actually fixes the issue, because AFAICT we do in fact have a need for reconnection in such cases.

**Comment by michaelfig** [comment]
> We need an issue to describe the problem so this PR can be marked as closing it. @michaelfig do you have a good explanation for that?

Yes.  I will provide an issue.

> Also, I think we're still missing test coverage of [... Alice] the vat watching a non-vow promise decided by [Bob] a remote vat [where Alice] is itself restarted before [Bob fulfils] that promise (e.g., `watch(E(remotable).getLongLivedPromise(), makeIncarnationCrossingDurableWatcher(…))`).

I agree we're missing coverage, and I will add corresponding tests.  I see now that `@agoric/boot` is the right place for such tests.

> And I'd like clarification from @iomekam that this actually fixes the issue, because AFAICT we do in fact have a need for reconnection in such cases.

I don't call this "promise reconnection", since the original promise is not disconnected.  Instead this is "watchers persisting across incarnations", which is a feature already provided by `watchPromise`.  The solution in this PR is to make sure that there is always a persistent watcher, even though the first watcher has no way to retry on rejection.  Subsequent watchers can retry only if the first watcher is fulfilled with a vow to persist and use the shortening protocol.

File: packages/vow/src/watch.js [hunk 6]
```diff
@@ -74,58 +67,50 @@ const prepareWatcherKit = (zone, rejectionMeansRetry, watchVow) =>
...
-            settle(resolver, watcher, 'onRejected', reason);
-          } else {
-            resolver.reject(reason);
-          }
-        },
+      /** @type {Required<import('./watch-promise.js').PromiseWatcher>['onRejected']} */
+      onRejected(reason) {
+        const { vow, watcher, resolver } = this.state;
+        if (vow && rejectionMeansRetry(reason)) {
+          watchVow(vow, this.self);
```
Comment (gibson042) [inline packages/vow/src/watch.js]: Which test covers this scenario, in which a watched promise resolves to a vow that is subsequently rejected with a reason recognized by `rejectionMeansRetry`, and what is this `watchVow` expected to accomplish in such a case? Because it seems to me like an infinite loop.

File: packages/vow/src/watch.js [hunk 7]
```diff
@@ -156,19 +145,11 @@ export const prepareWatch = (
...
-      // Persistently watch the specimen.
-      if (!payload) {
-        // Specimen is not a vow.
-        promiseWatcher.onFulfilled(specimen);
-        return;
-      }
-      watchVow(specimen, promiseWatcher);
-    }).catch(e => promiseWatcher.onRejected(e));
+    // Coerce the specimen to a promise, and start the watcher cycle.
+    watchPromise(basicE.resolve(specimenP), promiseWatcher);
```
Comment (gibson042) [inline packages/vow/src/watch.js:152]: Is this coercion appropriate? @warner: What happens to non-exported local promises when their containing vat is restarted?

Comment (michaelfig) [reply packages/vow/src/watch.js:152]: `watchPromise` adds the promises to liveslots's internal tables, so that they have an id that is reconstituted in the next incarnation.

File: packages/vow/src/watch.js (hunk 6 above)
Comment (michaelfig) [reply packages/vow/src/watch.js]: These are the tests, in `agoric-sdk/packages/vow/test/test-disconnect.js`.
```
    ℹ testing (plan=disco,happy, watchVow=true)
    ℹ testing (plan=disco,sad, watchVow=true)
    ℹ testing (plan=disco,sad,disco,happy, watchVow=true)
    ℹ testing (plan=disco,disco,happy, watchVow=true)
    ℹ testing (plan=disco,disco,sad, watchVow=true)
```
It's not an infinite loop because `watchVow` obtains a fresh promise from the vow internals's `shorten()` method.

File: packages/vow/src/vow-utils.js (hunk 1 above)
Comment (michaelfig) [reply packages/vow/src/vow-utils.js]: I ripped out `unwrapPromise` because it was only useful when I was trying to be defensive in a couple of places.  Turns out that making `when` an async function eliminated the need.

File: packages/vow/src/watch.js (hunk 2 above)
Comment (michaelfig) [reply packages/vow/src/watch.js]: That's correct.  I'll find a place to document it.

File: packages/vow/test/test-watch.js (hunk 4 above)
Comment (michaelfig) [reply packages/vow/test/test-watch.js]: Agreed.

File: packages/vow/test/test-watch.js (hunk 5 above)
Comment (michaelfig) [reply packages/vow/test/test-watch.js:106]: I renamed the test and added some comments.

File: packages/vow/src/watch.js (hunk 6 above)
Comment (michaelfig) [reply packages/vow/src/watch.js]: I'll rename `watchVow` to `retryAndWatchVow`.

File: packages/vow/src/watch.js (hunk 7 above)
Comment (gibson042) [reply packages/vow/src/watch.js:152]: Right, but code anything actually resolve them? I guess the point is moot, since `E.resolve(notAPromise)` is born fulfilled.

File: packages/vow/src/watch.js (hunk 3 above)
Comment (michaelfig) [reply packages/vow/src/watch.js]: I thought it was a mistake to allow `watchPromise` to use variadic arguments.  I am willing to accept a third argument with the same semantics as `watchPromise`'s third, but I will reserve `watch`'s fourth for an options bag if it ever becomes clear that we need options.

File: packages/vow/src/watch.js (hunk 6 above)
Comment (gibson042) [reply packages/vow/src/watch.js]: Ah, I see. So it would only be infinite with a fake vow whose `shorten` method maliciously returned disconnection messages.

File: packages/vow/src/watch.js (hunk 3 above)
Comment (michaelfig) [reply packages/vow/src/watch.js]: > can you explain why [it] should return a vow that resolves to results from the watcher

For parity with `Promise.prototype` methods like `.then`, and `.catch`'s chaining functionality, to allow composing vows built with watchers with other watchers.

File: packages/vow/src/watch.js (hunk 7 above)
Comment (michaelfig) [reply packages/vow/src/watch.js:152]: Looking through `agoric-sdk/packages/swingset-liveslots/src/watchedPromises.js`, it constructs a `loadWatchedPromiseTable` function.  ISTR tracing through this whole flow to find that if a local promise was watched, then the vat was restarted, it was managed correctly.  Something in liveslots would end up calling the `watcher.onRejected` methods for the prior incarnation's promises that were no longer reachable.

But that requires @warner or @mhofman to confirm.

Comment (gibson042) [reply packages/vow/src/watch.js:152]: Looks like your memory of what happens is correct: https://github.com/Agoric/agoric-sdk/blob/18d561f1b95755e58ca691b26938fc249618018d/packages/SwingSet/test/test-promises.js#L219-L292

But as for the mechanics, it seems to be driven by the kernel rather than liveslots: https://github.com/Agoric/agoric-sdk/blob/18d561f1b95755e58ca691b26938fc249618018d/packages/SwingSet/src/kernel/kernel.js#L902-L905

Albeit with liveslots cooperation in the form of exporting all watched promises: https://github.com/Agoric/agoric-sdk/blob/18d561f1b95755e58ca691b26938fc249618018d/packages/swingset-liveslots/src/watchedPromises.js#L223-L226

So basically, there's no such thing as a non-exported local watched promise.

File: packages/vow/src/tools.js [hunk 8]
```diff
@@ -11,15 +11,8 @@ import { prepareWatch } from './watch.js';
  */
 export const prepareVowTools = (zone, powers = {}) => {
   const { rejectionMeansRetry = () => false, watchPromise } = powers;
```
Comment (gibson042) [inline packages/vow/src/tools.js]: The `rejectionMeansRetry` name has seemed a little off to me... thoughts on renaming it to something like `isRetryableRejection` before a release includes this package?

File: packages/vow/src/watch.js [hunk 9]
```diff
@@ -1,15 +1,13 @@
...
 const { apply } = Reflect;
 
 /**
  * @param {typeof watchPromiseShim} watchPromise
  */
-const makeWatchVow =
+const makeRetryAndWatchVow =
   watchPromise =>
   /**
    * @param {any} specimen
```
Comment (gibson042) [inline packages/vow/src/watch.js]:
```suggestion
  /**
   * Associate a (usually durable) watcher object with a vow or promise.
   *
   * @param {any} specimen
```
(at which point we can probably go back to `watchVow` rather than `retryAndWatchVow`)

File: packages/boot/test/upgrading/vat-vow.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  /** @type {PublishKit<null>} */
+  const { publisher, subscriber } = makePublishKit();
+  publisher.publish(null);
+
+  let numRejected = 0n;
+  let numFulfilled = 0n;
+  const makeWatcher = zone.exoClass(
+    'PromiseWatcher',
+    undefined,
+    (expectFulfilled, value) => ({ expectFulfilled, value }),
```
Comment (gibson042) [inline packages/boot/test/upgrading/vat-vow.js]: These parameters and fields appear to be unused.

File: packages/boot/test/upgrading/vat-vow.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+      watch(Promise.resolve('hello'), makeWatcher(true, 'hello'));
+
+      for await (const _ of subscribeLatest(subscriber)) {
+        if (numRejected > 0n) {
+          throw new Error('unexpected rejections');
+        }
+        if (numFulfilled > 0n) {
+          break;
+        }
+      }
```
Comment (gibson042) [inline packages/boot/test/upgrading/vat-vow.js]: I find this style to be very opaque; can we instead put the test logic in packages/boot/test/upgrading/test-upgrade-vats.js?
```js
  const vowRoot = await EV.vat('bootstrap').createVat(vowVatConfig);

  t.log('test incarnation 0');
  /** @type {Record<string, [settlementValue?: unknown, isRejection?: boolean]>} */
  const localPromises = { forever: [], fulfilled: ['hello'] };
  await EV(vowRoot).makeLocalPromiseWatchers(localPromises);
  t.deepEqual(await EV(vowRoot).getWatcherResults(), {
    fulfilled: { status: 'fulfilled', value: 'hello' },
    forever: { status: 'unsettled' },
  });

  t.log('restart');
  const { incarnationNumber } =
    await EV.vat('bootstrap').upgradeVat(vowVatConfig);
  t.is(incarnationNumber, 1, 'vat must be reincarnated');

  t.log('test incarnation 1');
  t.deepEqual(await EV(vowRoot).getWatcherResults(), {
    fulfilled: { status: 'fulfilled', value: 'hello' },
    forever: {
      status: 'rejected',
      reason: {
        name: 'vatUpgraded',
        upgradeMessage: 'vat upgraded',
        incarnationNumber: 0,
      },
    },
  });
```

File: packages/vow/src/watch.js (hunk 3 above)
Comment (gibson042) [reply packages/vow/src/watch.js]: > > can you explain why [it] should return a vow that resolves to results from the watcher
> 
> For parity with `Promise.prototype` methods like `.then`, and `.catch`'s chaining functionality, to allow composing vows built with watchers with other watchers.

Do we have a use case for that? We've gotten this far without anything similar in the `watchPromise` provided by liveslots, so I'm inclined to YAGNI it out of vow `watch` as well—especially since there are likely to be subtle differences/complexities (e.g., what happens if an `onFulfilled` method returns the vow returned from the `watch` to which its object was passed?[^1]) and because nothing about names like "onFulfilled" and "onRejected" imply that the return value is relevant.

[^1]: Spoiler: an infinite loop, in contrast to errors from analogous situations with native promises and with `E`.

File: packages/vow/src/watch.js [hunk 12]
```diff
@@ -142,33 +117,35 @@ export const prepareWatch = (
...
   /**
    * @template [T=any]
    * @template [TResult1=T]
    * @template [TResult2=T]
    * @param {import('./types.js').ERef<T | import('./types.js').Vow<T>>} specimenP
    * @param {import('./types.js').Watcher<T, TResult1, TResult2>} [watcher]
+   * @param {unknown} [watcherContext]
    */
-  const watch = (specimenP, watcher) => {
+  const watch = (specimenP, watcher, watcherContext) => {
```
Comment (gibson042) [inline packages/vow/src/watch.js:139]: We already have a case of calling `watchPromise` with more than one context argument: https://github.com/Agoric/agoric-sdk/blob/18d561f1b95755e58ca691b26938fc249618018d/packages/zoe/src/zoeService/startInstance.js#L100-L102

Such patterns _could_ be converted from `watchPromise(awaitable, watcher, instAdmin, adminNode)` to `watchPromise(awaitable, watcher, [instAdmin, adminNode])` or similar, but that's likely to cost brevity with ESLint and also to be disruptive of the watcher code for transitioning. Do you have a strong reason to prefer a singular `watcherContext` over variadic `...callbackArgs`?

File: packages/vow/src/when.js [hunk 13]
```diff
@@ -1,51 +1,39 @@
...
+   * Shorten any vows in `specimenP` until we achieve a final result.
+   *
    * @template T
    * @param {import('./types.js').ERef<T | import('./types.js').Vow<T>>} specimenP
    */
-  const when = specimenP => {
-    /** @type {import('./types.js').VowKit<T>} */
-    const { resolver } = makeVowKit();
-    const promise = providePromiseForVowResolver(resolver);
+  const when = async specimenP => {
```
Comment (gibson042) [inline packages/vow/src/when.js]: Another interface consistency question: shouldn't vow `when` have the same (awaitable, onFulfilled, onRejected) parameter signature as [`E.when`](https://github.com/endojs/endo/blob/7639f0ceb6f23dfdae5c925962c17a9368bac760/packages/eventual-send/src/E.js#L222-L237), rather than just (awaitable) like `Promise.resolve`?

File: packages/vow/test/test-disconnect.js [hunk 14]
```diff
@@ -52,18 +52,18 @@ test('retry on disconnection', async t => {
     [2, 'disco', 'disco', 'sad'],
   ];
 
-  for await (const watchVow of [false, true]) {
-    t.log('testing watchVow', watchVow);
+  for await (const doWatch of [false, true]) {
+    t.log('testing watch', doWatch);
```
Comment (gibson042) [inline packages/vow/test/test-disconnect.js]: Suggestion: replace these boolean values with descriptive strings.
```suggestion
  for await (const pattern of ['when', 'watch']) {
    t.log('testing', pattern);
```

File: packages/vow/test/test-disconnect.js [hunk 15]
```diff
@@ -52,18 +52,18 @@ test('retry on disconnection', async t => {
...
 
       /** @type {import('../src/types.js').Vow<string>} */
       const vow = makeTagged('Vow', {
         vowV0: makeTestVowV0(plan),
       });
 
       let resultP;
-      if (watchVow) {
+      if (doWatch) {
         const resultW = watch(vow, {
```
Comment (gibson042) [inline packages/vow/test/test-disconnect.js]: What is the purpose of the `t.is('then' in resultW, false, 'watch resultW.then is undefined')` a dozen lines below this? Is it important that `resultW` is _not_ thenable, or that it _is_ a vow?

File: packages/vow/test/test-watch.js [hunk 16]
```diff
@@ -10,46 +12,99 @@ import { prepareVowTools } from '../src/tools.js';
...
+  t.is(
+    await when(
+      // @ts-expect-error intentional extra argument
+      watch(connVow3P, makeAckWatcher(packet), 'watcher context', 'unexpected'),
+    ),
+    'rejected',
+  );
+});
+
+test('disconnection of non-vow informs watcher', async t => {
```
Comment (gibson042) [inline packages/vow/test/test-watch.js:85]: :+1:

File: packages/vow/test/test-watch.js [hunk 17]
```diff
@@ -10,46 +12,99 @@ import { prepareVowTools } from '../src/tools.js';
...
+  const { watch, when } = prepareVowTools(zone, {
+    rejectionMeansRetry: reason => reason === 'disconnected',
+  });
+
+  // Even though this promise is rejected with a retry reason, there's no vow
+  // before it to retry, so we pass the rejection up to the watcher.
+  const { reject, promise } = makePromiseKit();
+  reject('disconnected');
+
+  const vow = watch(promise, {
```
Comment (gibson042) [inline packages/vow/test/test-watch.js]:
```suggestion
  // Even though this promise is rejected with a retryable reason, there's no vow
  // before it to retry, so we pass the rejection up to the watcher.
  const vow = watch(Promise.reject('disconnected'), {
```

**Review by gibson042 — APPROVED** [review]
I have some reservations about the vow API surface area (comments below, in addition to https://github.com/Agoric/agoric-sdk/pull/8992#discussion_r1506420203 ), but those are independent of #9002. Please fix here anything that you feel is appropriate and reply to everything else for potential followup issues.

File: packages/boot/test/upgrading/vat-vow.js (hunk 10 above)
Comment (michaelfig) [reply packages/boot/test/upgrading/vat-vow.js]: Vestigial.  I'll remove it when rewriting the test.

File: packages/vow/src/tools.js (hunk 8 above)
Comment (michaelfig) [reply packages/vow/src/tools.js]: Even though this name has landed on master, we can gamble that no external consumer is actually using this config option.  I'll make the change.

File: packages/vow/src/watch.js (hunk 9 above)
Comment (michaelfig) [reply packages/vow/src/watch.js]: I'll call it `watchNextStep`.

File: packages/vow/src/watch.js (hunk 12 above)
Comment (michaelfig) [reply packages/vow/src/watch.js:139]: My reason is that all of the watcher callback argument positions are exhausted, as well as all of the `watch` argument positions.  I think that is bad design because it paints us into a corner, for no good reason.

I'm not sure what you mean by "cost brevity with ESLint": does it matter that converting from `watchPromise` to `watch` already saves you 7 letters?  The typing ensures you see red squigglies for the extra arguments, both on `watch` and the watcher callbacks themselves.

File: packages/vow/src/when.js (hunk 13 above)
Comment (michaelfig) [reply packages/vow/src/when.js]: That's a reasonable question, mainly decided by the evolution of the API over time.  I can certainly accommodate it.

File: packages/vow/test/test-disconnect.js (hunk 15 above)
Comment (michaelfig) [reply packages/vow/test/test-disconnect.js]: It must *not* be a thenable, or `@agoric/marshal` will choke on it.

**Comment by michaelfig** [comment]
> We've gotten this far without anything similar in the `watchPromise` provided by liveslots, so I'm inclined to YAGNI it out of vow `watch` as well

IAGNI... in the network vat, we have patterns that look like this:

```js
const result = E(obj).method(...args).then(r => processResult(r)).catch(e => handleRejection(e))
// in consumer
result.then(r => dispatchResult(r));
```

Without chaining, we'd need to create an amalgam of all of those handlers to use as a single watcher.  That seems tricky, especially when those operations are trusted differently.  With chaining, we get:

```js
const processed = watch(E(obj).method(...args), resultProcessingWatcher);
const result = watch(processed, rejectionHandlingWatcher);
// in consumer
watch(result, resultDispatchingWatcher);
```

Handling an infinite vow loop is a different matter... you can create one even without chaining.

File: packages/vow/src/watch.js (hunk 3 above)
Comment (michaelfig) [reply packages/vow/src/watch.js]: > We've gotten this far without anything similar in the `watchPromise` provided by liveslots, so I'm inclined to YAGNI it out of vow `watch` as well

IAGNI... in the network vat, we have patterns that look like this:

```js
const result = E(obj).method(...args).then(r => processResult(r)).catch(e => handleRejection(e))
// in consumer
result.then(r => dispatchResult(r));
```

Without chaining, we'd need to create an amalgam of all of those handlers to use as a single watcher.  That seems tricky, especially when those operations are trusted differently.  With chaining, we get:

```js
const processed = watch(E(obj).method(...args), resultProcessingWatcher);
const result = watch(processed, rejectionHandlingWatcher);
// in consumer
watch(result, resultDispatchingWatcher);
```

Handling an infinite vow loop is a different matter... you can create one even without chaining.

File: packages/vow/src/watch.js (hunk 12 above)
Comment (gibson042) [reply packages/vow/src/watch.js:139]: > I'm not sure what you mean by "cost brevity with ESLint": does it matter that converting from `watchPromise` to `watch` already saves you 7 letters?

I'm thinking of situations I've encountered where it insists upon line breaks. But that's admittedly not a huge deal either... any signature can work, but you're right that claiming all arguments is a risk. There's just a tradeoff with compatibility. Given the uncertainty, let's leave it for now.
